pkg/plugin/plugin.go

```diff
@@ -65,66 +66,127 @@ func (p *Plugin) SetInformation(info *Information) {
 }
 
 func (p *Plugin) Status() Status {
-	p.mutex.RLock()
-	defer p.mutex.RUnlock()
-
 	return p.status
 }
 
 func (p *Plugin) Kill() error {
 	p.mutex.Lock()
+	defer p.mutex.Unlock()
 
 	p.status = ShuttingDown
 
 	err := p.command.Process.Kill()
 	if err != nil {
 		p.status = Crashed
-		p.mutex.Unlock()
 
 		return fmt.Errorf("killing process: %w", err)
 	}
 
-	p.mutex.Unlock()
-
 	err = (*p.command).Wait()
 	if err.Error() != "signal: killed" {
-		p.mutex.Lock()
 		p.status = Crashed
-		p.mutex.Unlock()
 
 		return fmt.Errorf("killing process: unexpected wait error: got %w, want `signal: killed`", err)
 	}
 
+	p.status = Down
+
 	return nil
 }
 
 func (p *Plugin) ReverseProxy() *httputil.ReverseProxy {
 	return p.reverseProxy
 }
 
-func New(path string, id int64) (*Plugin, error) {
-	//nolint:exhaustruct
-	plgn := &Plugin{status: Starting}
+type prefixWriter struct {
+	prefix string
+}
+
+func (w *prefixWriter) Write(p []byte) (n int, err error) {
+	//nolint:forbidigo,wrapcheck
+	return fmt.Print(w.prefix + string(p))
+}
+
+func (p *Plugin) Start() error {
+	pluginName := filepath.Base(p.BinPath)
+
+	log.Printf("Starting plugin '%s' with id %d\n", pluginName, p.ID)
+
+	p.mutex.Lock()
 
-	plgn.command = exec.Command(path, strconv.FormatInt(id, 10)) // #nosec G204
+	status := p.Status()
 
-	pipe, err := plgn.command.StdoutPipe()
+	if status != Down && status != Starting {
+		return fmt.Errorf("plugin is not ready to start")
+	}
+
+	p.command = exec.Command(p.BinPath, strconv.FormatInt(p.ID, 10)) // #nosec G204
+
+	stdOut, err := p.command.StdoutPipe()
 	if err != nil {
-		return nil, fmt.Errorf("instantiating plugin %s: initializing stdout Pipe: %w", path, err)
+		return fmt.Errorf("start plugin %s: initializing stdout Pipe: %w", pluginName, err)
 	}
 
-	plgn.stdOut = pipe
+	stdErr, err := p.command.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("start plugin %s: initializing stderr Pipe: %w", pluginName, err)
+	}
 
-	pipe, err = plgn.command.StderrPipe()
+	err = p.command.Start()
 	if err != nil {
-		return nil, fmt.Errorf("instantiating plugin %s: initializing stderr Pipe: %w", path, err)
+		return fmt.Errorf("start plugin %s: starting command: %w", pluginName, err)
 	}
 
-	plgn.stdErr = pipe
+	p.status = Up
+
+	p.mutex.Unlock()
+
+	prefixedStdout := io.MultiWriter(os.Stdout, &prefixWriter{prefix: fmt.Sprintf("[%s] - ", pluginName)})
+
+	// start two goroutine for stdout and stderr
+	//nolint:errcheck
+	go io.Copy(prefixedStdout, stdOut)
+	//nolint:errcheck
+	go io.Copy(os.Stderr, stdErr)
+
+	// start a goroutine to wait on the command
+	go func() {
+		//nolint:errcheck
+		p.command.Wait()
+		log.Printf("plugin '%s' exiting.\n", pluginName)
+
+		p.mutex.Lock()
+		defer p.mutex.Unlock()
+		p.status = Down
```

Comment:
Down or Crashed ? You should have a look at the error returned by p.command.Wait()